server/api_archive.go

```diff
@@ -0,0 +1,178 @@
+These files all contain publicly available information
+and additional work is needed to stand up a new server
+using them.
+`
+
+// Zip archive helper.
+type zipArchiveWriter struct {
+	buffer *bytes.Buffer
+	znw    *zip.Writer
+}
+
+// Create a new zip archive writer.
+func NewArchive() *zipArchiveWriter {
+	buf := new(bytes.Buffer)
+	return &zipArchiveWriter{
+		buffer: buf,
+		znw:    zip.NewWriter(buf),
+	}
+}
+
+// Close the archive writer and return a buffer to its contents.
+func (arc *zipArchiveWriter) Close() (*bytes.Buffer, error) {
+	if err := arc.znw.Close(); err != nil {
+		return nil, nil
+	}
+	return arc.buffer, nil
+}
+
+func (arc *zipArchiveWriter) AddFile(reader io.Reader, name string, modTime time.Time) error {
+	writer, err := arc.znw.CreateHeader(&zip.FileHeader{
+		Name:     name,
+		Method:   zip.Deflate,
+		Modified: modTime,
+	})
+	if err != nil {
+		return err
+	}
+	if _, err := io.Copy(writer, reader); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ArchiveHandler provides the GET /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+	if !gcas.ApiArchiveRateLimiter.Allow() {
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+	arc := NewArchive()
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only while happening in a single write. We must archive the files in reverse order from
+	// their dependency order.
+	for _, name := range PublicFiles {
+		err := gcas.addFile(name, arc)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+	// Add the pseudo file server.pubkey last, as the other files are signed by it.
+	if err := gcas.addPubKeyFile("server.pubkey", arc); err != nil {
+		http.Error(w, fmt.Sprintf("Error zipping server.pubkey: %v", err), http.StatusInternalServerError)
+		return
+	}
+	if err := gcas.addReadmeFile(arc); err != nil {
+		http.Error(w, fmt.Sprintf("Error adding README: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Close archive and write out the response
+	buf, err := arc.Close()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("%v", err), http.StatusInternalServerError)
+	}
+	w.Header().Set("Content-Type", "application/zip")
+	if _, err := w.Write(buf.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
```

Comment:
need to return after calling http.Error. I know this is the end of the function, but it's still  best practice in case code gets shuffled around later

Reply:
Agree. The problem was cut and paste from the earlier block that did not have a return :)